ssl/statem/statem_srvr.c

```diff
@@ -3513,6 +3525,98 @@ WORK_STATE tls_post_process_client_key_exchange(SSL_CONNECTION *s,
     return WORK_FINISHED_CONTINUE;
 }
 
+MSG_PROCESS_RETURN tls_process_client_rpk(SSL_CONNECTION *sc, PACKET *pkt)
+{
+    MSG_PROCESS_RETURN ret = MSG_PROCESS_ERROR;
+    SSL_SESSION *new_sess = NULL;
+    EVP_PKEY *peer_rpk = NULL;
+
+    /*
+     * To get this far we must have read encrypted data from the client. We no
+     * longer tolerate unencrypted alerts. This is ignored if less than TLSv1.3
+     */
+    if (sc->rlayer.rrlmethod->set_plain_alerts != NULL)
+        sc->rlayer.rrlmethod->set_plain_alerts(sc->rlayer.rrl, 0);
+
+    if (!tls_process_rpk(sc, pkt, &peer_rpk)) {
+        /* SSLfatal already called */
+        goto err;
+    }
+
+    if (peer_rpk == NULL) {
```

Comment:
Is this a possible state? If the client sends neither a certificate chain, nor an RPK, how did we know to expect an RPK?  Is it based in the client's supported certificate type list alone?

Reply:
There are two points:

- regardless if it's x509 or RPK, the first to read is a 3 bytes length. If that is 0, it's empty.

- considering RFC7250

```
   opaque ASN.1Cert<1..2^24-1>;

   struct {
       select(certificate_type){

            // certificate type defined in this document.
            case RawPublicKey:
              opaque ASN.1_subjectPublicKeyInfo<1..2^24-1>;

           // X.509 certificate defined in [RFC 5246](https://www.rfc-editor.org/rfc/rfc5246)
           case X.509:
             ASN.1Cert certificate_list<0..2^24-1>;

           // Additional certificate type based on
           // "TLS Certificate Types" subregistry
       };
   } Certificate;
```

A certificate has at least 1 byte, while a list may have 0 bytes. That doesn't make a real difference, as both would be just a 0 in 3 bytes.

Reply:
You're mistaken. If the claimed length of the RPK is 0, it is a protocol violation, the minimum RPK length is 1 byte, that would be encoded as 1 byte *after* the 3 byte length, an example would be:
```
 00 00 01 XX
```
Of course there is no such SPKI value since it is an ASN.1 sequence of object ID, possible parameters and BIT string, so would be at least a bunch of bytes longer.  The `1` is just to make clear that it can't be empty.

Reply:
> You're mistaken. 

Isn't the point more, what you want to achieve? A empty certificate is used, if a peer has no matching certificate.

We can now stick to the "words" in RFC7250, assuming, that this case of "empty certificates" was well considered and then put into this "1" instead of explicit stating, that "empty certificates must not be supported"? I don't think so. Such an RFC would be more a miracle than help.

We can also go for [RFC 5246 7.4.6](https://www.rfc-editor.org/rfc/rfc5246#section-7.4.6):

> If no suitable certificate is available, the client MUST send a certificate message containing no certificates.  > That is, the certificate_list structure has a length of zero. 

With that, if you have no suitable RPK certificate, you MUST send a empty list.

Should we go to ask Hannes, what he and the other authors had in mind for "empty rpk certificates"?

Reply:
So my proposed patch is:
```c
--- a/ssl/statem/statem_lib.c
+++ b/ssl/statem/statem_lib.c
@@ -1056,3 +1056,3 @@ int tls_process_rpk(SSL_CONNECTION *sc, PACKET *pkt, EVP_PKEY **peer_rpk)
     PACKET context;
-    unsigned long cert_list_len, spki_len = 0;
+    unsigned long spki_len = 0;
     const unsigned char *spki, *spkistart;
@@ -1085,9 +1085,3 @@ int tls_process_rpk(SSL_CONNECTION *sc, PACKET *pkt, EVP_PKEY **peer_rpk)
 
-    if (!PACKET_get_net_3(pkt, &cert_list_len)
-            || PACKET_remaining(pkt) != cert_list_len) {
-        SSLfatal(sc, SSL_AD_DECODE_ERROR, SSL_R_LENGTH_MISMATCH);
-        goto err;
-    }
-
-    /* RPK only admits a single CertificateEntry in the list, so no loop. */
+    /* RPK consists of a single optional non-empty entry, not a list. */
     if (!PACKET_get_net_3(pkt, &spki_len)) {
@@ -1206,7 +1200,2 @@ unsigned long tls_output_rpk(SSL_CONNECTION *sc, WPACKET *pkt, CERT_PKEY *cpk)
 {
-    if (!WPACKET_start_sub_packet_u24(pkt)) {
-        SSLfatal(sc, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
-        return 0;
-    }
-
     if (!tls_add_rpk(sc, pkt, cpk)) {
@@ -1215,8 +1204,2 @@ unsigned long tls_output_rpk(SSL_CONNECTION *sc, WPACKET *pkt, CERT_PKEY *cpk)
     }
-
-    if (!WPACKET_close(pkt)) {
-        SSLfatal(sc, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
-        return 0;
-    }
-
     return 1;
```

Reply:
> The loophole if you like is that a zero length signals an empty certificate list, not an empty RPK. 

FMPOV, the term "empty list" mainly reflects, that it's encoded with 0x00 00 00 and signals, that there is no matching certificate.

The way, how that is implemented, may differ. One may decide to first read the length (regardless of RPK or certificate list) and then decide to handle 0, and on lengths > 0, to go for the certificate list or rpk.
Or to go for the certificate list or rpk, which then requires to handle a 0 length for rpk, what may be considered to be "wrong according RFC 7250", but in the end it's just a missing client certificate.

Once the implementation is stable about the missing client certificate, let me know, I will extend my interoperability tests.

Reply:
FWIW, my temporary fork of the PR (which enables me to test the Postfix implementation, making for a more thorough review) has what I believe to be the final form of certificate message codec.  This is at https://github.com/vdukhovni/openssl in the "rpk-fixup" branch: https://github.com/vdukhovni/openssl/commits/rpk-fixup

Reply:
I don't believe this takes into account TLSv1.3 certificate extensions. The extra length wrapper is needed for that. You are correct for TLSv1.2, however.

Reply:
That patch is outdated, the "rpk-fixup" branch in my "fork" is the correct version.

Reply:
I haven't had the chance to look at that branch yet.